### ISSUE TITLE
scMultiSim

### ISSUE THREAD

**Issue by lhc70000** [Bioconductor/Contributions#3068]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/ZhangLabGT/scMultiSim

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand Bioconductor [Package Naming Policy][9] and acknowledge
  Bioconductor may retain use of package name.

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.
  
- [x] I am familiar with the [Bioconductor code of conduct][7] and 
  agree to abide by it.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For questions/help about the submission process, including questions about
the output of the automatic reports generated by the SPB (Single Package
Builder), please use the #package-submission channel of our Community Slack.
Follow the link on the home page of the [Bioconductor website][8] to sign up.

[1]: https://contributions.bioconductor.org/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/
[7]: https://bioconductor.org/about/code-of-conduct/
[8]: https://bioconductor.org/
[9]: https://bioconductor.org/developers/package-submission/#naming

**Comment by vjcitn** [comment]
```
* checking re-building of vignette outputs ... ERROR
Error(s) in re-building vignettes:
  ...
--- re-building ‘basics.Rmd’ using knitr
Error: processing vignette 'basics.Rmd' failed with diagnostics:
there is no package called ‘markdown’
--- failed re-building ‘basics.Rmd’

--- re-building ‘spatialCCI.rmd’ using knitr

Quitting from lines 75-94 [unnamed-chunk-1] (spatialCCI.rmd)
Error: processing vignette 'spatialCCI.rmd' failed with diagnostics:
object 'GRN_params_100' not found
--- failed re-building ‘spatialCCI.rmd’

SUMMARY: processing the following files failed:
  ‘basics.Rmd’ ‘spatialCCI.rmd’

Error: Vignette re-building failed.
Execution halted
```
if you need markdown, please declare in DESCRIPTION.  fix the data availability error.  thanks

**Comment by lhc70000** [comment]
Thank you very much. I have updated the repo to fix these two problems.

**Comment by ococrook** [comment]
Hi, I'm looking forward to reviewing the package! Please fix the build errors before I can proceed.

**Comment by lhc70000** [comment]
Hi @ococrook! I'm so sorry for the delay. I wanted to resolve the timeout issue; however, since scMultiSim is a data simulator, some simulation code in the vignette may take a long time to run (~10 minutes). I'm considering either
- reduce the number of cells/genes simulated in the vignette
- include the demo data with the package and load it directly into the vignette

Which one do you prefer, or is there any other solution you would recommend? Thanks!

**Comment by ococrook** [comment]
Hi!

How big the demo data - this maybe your limitation if it over 5mb you would have to store it in a data package. I favour the first option if it makes sense to do that - and allow your user to locally increase the number of cells and genes easily.

**Comment by lhc70000** [comment]
Hi @ococrook, sorry for the delay! Now I'll prioritize this package and all build errors have been fixed. I will also look into the warnings soon.

I'm committed to promptly addressing any review comments. Thank you for your time with this large package!

**Comment by lhc70000** [comment]
Hi @ococrook! The package now builds without errors or warnings. I apologize once more for the delay caused by my personal issues. I look forward to your review and will promptly address any comments.

**Comment by ococrook** [comment]
Thanks! Sorry I didnt see the notification. I'll get to this shortly

**Comment by ococrook** [comment]
This is great! A few things to tidy up:

- [ ] The News.md version doesnt match the version of the package, remember to bump the package version and the news version by sub increments
- [ ] You have a lot of imports - are they all necessary
- [ ] You have contrust like 1:N, which I tend to dislike can they be replaced by seq.int or seq_along?
- [ ] add bioc install instruction
- [ ] Could you add 1-line descriptors for all the helper functions it was hard to follow what they all did (especially important if theyre not going to exposed to the user and therefore documented in the manual files)

**Comment by ococrook** [comment]
Are you able to fix the timeout?

**Comment by lhc70000** [comment]
The problem this time was not the timeout…… My unit test failed on BioConductor's build server after I pushed the new changes.

```
Running the tests in tests/testthat.R failed.
Last 20 lines of output:
  
    `actual[3:15]`: 18 33 88 0 2 0 24 72 6  7 and 3 more...
  `expected[3:15]`: 18 33 88 5 4 0  8 96 0 18           ...
   Failure ('test-1_main.R:151:3'): simulates spatial data with discrete population and HGE 
  res$counts[selectedIndicies] (`actual`) not equal to c(...) (`expected`).
  
       actual  | expected               
   [1] 116.691 - 109.069  [1]           
   [2] 59.994  - 60.415   [2]           
   [3] 85.398  - 91.912   [3]           
   [4] 1.082   - 0.582    [4]           
   [5] 172.034 - 177.674  [5]           
   [6] 202.649 - 197.066  [6]           
   [7] 95.052  - 102.315  [7]           
   [8] 56.943  - 65.998   [8]           
   [9] 96.683  - 89.161   [9]           
  [10] 2.693   - 3.173    [10]          
   ... ...       ...      and 7 more ...
  
  [ FAIL 7 | WARN 0 | SKIP 0 | PASS 5 ]
  Error: Test failures
```

This is really strange since these tests successfully pass on my macOS and Linux machines. I'm still trying to figure it out; do you have any thoughts?

**Comment by ococrook** [comment]
Odd, usually a package dependcy or perhaps random seed issue?

**Comment by lhc70000** [comment]
Hi @ococrook, I have resolved the review comments and passed the Bioc Build System. Could you have a look? Thanks!

**Comment by ococrook** [comment]
@lshep where are the old build reports contained? I think we can accept this one shortly after double check.

**Comment by lshep** [comment]
http://staging.bioconductor.org:8000/jobs/1917/

**Comment by ococrook** [comment]
Sorry for the delay on my end - working only part time at the moment.

This is a very clear package as a I noted before.

Im keen to see some of the easy Notes cleared up e.g. vapply, 1:N constructions etc, corrected R dependency.

Make sure to run Biochcheck.  A few notes are fine but 13 is too many.

I realise not every package can depend on bioc infrastructure but could you give me a breif reason why this package does not depend on any bioconductor packages and why its more appropraite for the package on Bioconductor rather than CRAN?

**Comment by lhc70000** [comment]
Hi @ococrook,

Thank you for your feedback! I have reduced the number of NOTES to 6. Regarding the other NOTEs, such as the use of `<<-`, is due to the RefClasses in the package. Given the current structure and dependencies, altering this one might be challenging in the short term. Please tell me if you need to address other NOTEs.

Furthermore, for Bioconductor dependencies, I have added an option to output the simulated dataset as `SummarizedExperiment`. Additionally, in a separate work-in-progress branch, I am exploring the integration of `BiocParallel` to enhance performance (but that would be a future update). I am open to further suggestions and would welcome any additional recommendations on Bioconductor components that could be incorporated in this package.

**Comment by lhc70000** [comment]
Hi @ococrook, is there any update?

**Comment by ococrook** [comment]
Hi @lhc70000 

I thought i doubled checked this already - let me check now but that should be everything!

**Comment by ococrook** [comment]
HI @lhc70000 

Which branch are you pushing to on Github, is it main, develop or bioconductor - which one is being pushed to the bioconductor git just so I can double check the changes?

**Comment by lhc70000** [comment]
Sorry that I forgot to push some latest changes to GitHub. Just pushed to the `bioconductor` branch.

(Please note that I previously disabled some unit tests, because some changes in the latest Bioconductor build system caused floating point errors, and the accumulated errors made my test fail. I currently don't have a better solution for it, so I disabled them. I'm running the tests locally on macOS.)

**Comment by lhc70000** [comment]
Hi @ococrook, is there any update?

**Comment by lshep** [comment]
The default branch of your GitHub repository has been added to Bioconductor's
git repository as branch devel.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/lhc70000.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `BiocManager::install("scMultiSim")`. The package 'landing page' will be created at

https://bioconductor.org/packages/scMultiSim

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://git.bioconductor.org/BiocCredentials/
